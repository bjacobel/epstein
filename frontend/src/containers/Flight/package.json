{
  "main": "./Flight.js",
  "private": true
}
