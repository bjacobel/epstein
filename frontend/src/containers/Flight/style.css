@value margin from '../../stylesheets/measurements';
@value darkgrey from '../../stylesheets/colors';

.passengers {
  ul {
    margin: 0;
    padding-left: margin;
  }
}

.noslug {
  composes: robotoMono from '../../stylesheets/shared';
  color: darkgrey;
}
