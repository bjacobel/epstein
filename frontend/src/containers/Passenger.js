import React from 'react';
import { useQuery, gql } from '@apollo/client';

import Details from '../components/Details';
import PaginatedBrowser from '../components/PaginatedBrowser';
import Loading from '../components/Loading';
import MiniFlight from '../components/Mini/MiniFlight';
import { link } from '../stylesheets/shared.css';

const FLIGHT_LIMIT = 10;

const PASSENGER = gql`
  query Passenger($slug: String!, $offset: Int, $limit: Int!) {
    passenger(slug: $slug) {
      id
      name
      biography
      wikipedia_link
      image
      flightCount
      flights(offset: $offset, limit: $limit) {
        edges {
          id
        }
        pageInfo {
          count
          hasNext
        }
      }
    }
  }
`;

const range = histogram => {
  if (!histogram) return '';

  const months = Object.keys(histogram);
  const [firstDate] = months[0];
  const [lastDate] = months[months.length - 1];

  if (firstDate === lastDate) {
    return `in ${firstDate}`;
  } else {
    return `between ${firstDate} and ${lastDate}`;
  }
};

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data, fetchMore } = useQuery(PASSENGER, {
    variables: { slug, limit: FLIGHT_LIMIT },
  });

  if (loading) return <Loading text />;
  if (error) return <p>Error :(</p>;

  return (
    <Details>
      <span>name</span>
      <span>{data.passenger.name}</span>
      <span>bio</span>
      <div>
        <span>{data.passenger.biography}</span>
        <br />
        <a className={link} href={data.passenger.wikipedia_link}>
          <span>See more on Wikipedia â†’</span>
        </a>
      </div>
      <span>history</span>
      <div>
        <span>{`Appears in the flight logs at least ${data.passenger.flightCount} times`}</span>
        <span>{range(data.passenger.histogram)}</span>
      </div>
      {data.passenger.image ? (
        <>
          <span>image</span>
          <img src={data.passenger.image} alt="" />
        </>
      ) : null}
      <span>flights</span>
      <PaginatedBrowser
        browserComponent={MiniFlight}
        ids={data.passenger.flights.edges.map(x => x.id)}
        totalAvailable={data.passenger.flightCount}
        pageSize={FLIGHT_LIMIT}
        fetchMore={() =>
          fetchMore({
            variables: { offset: data.passenger.flights.edges.length + 1 },
            updateQuery: (prev, { fetchMoreResult }) => {
              if (!fetchMoreResult) return prev;
              return {
                passenger: {
                  ...prev.passenger,
                  flights: {
                    ...fetchMoreResult.passenger.flights,
                    pageInfo: {
                      ...fetchMoreResult.passenger.flights.pageInfo,
                      count:
                        prev.passenger.flights.pageInfo.count +
                        fetchMoreResult.passenger.flights.pageInfo.count,
                    },
                    edges: [
                      ...prev.passenger.flights.edges,
                      ...fetchMoreResult.passenger.flights.edges,
                    ],
                  },
                },
              };
            },
          })
        } // eslint-disable-line
      />
    </Details>
  );
};
