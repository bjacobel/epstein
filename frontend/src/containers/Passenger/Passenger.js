import React from 'react';
import { useQuery, gql } from '@apollo/client';
import { format, parseISO, isEqual, isSameYear } from 'date-fns';

import Details from '../../components/Details';
import PaginatedBrowser from '../../components/PaginatedBrowser';
import Loading from '../../components/Loading';
import MiniFlight from '../../components/Mini/MiniFlight';
import { link } from '../../stylesheets/shared.css';
import MetaTags from '../../components/MetaTags';

const FLIGHT_LIMIT = 10;

export const PASSENGER = gql`
  query Passenger($slug: String!, $offset: Int, $limit: Int!) {
    passenger(slug: $slug) {
      id
      name
      slug
      biography
      wikipedia_link
      image
      flightCount
      histogram {
        count
        month
      }
      flights(offset: $offset, limit: $limit) {
        edges {
          id
        }
        pageInfo {
          count
          hasNext
        }
      }
    }
  }
`;

export const range = histogram => {
  if (!histogram.length) return '';
  const firstDate = parseISO(histogram[0].month);
  const lastDate = parseISO(histogram[histogram.length - 1].month);

  const firstFormatted = format(firstDate, 'MMMM y');
  const lastFormatted = format(lastDate, 'MMMM y');

  if (isEqual(firstDate, lastDate)) {
    return `in ${firstFormatted}`;
  } else if (isSameYear(firstDate, lastDate)) {
    return `between ${format(firstDate, 'MMMM')} and ${lastFormatted}`;
  } else {
    return `between ${firstFormatted} and ${lastFormatted}`;
  }
};

const historyText = ({ passenger }) => {
  return `Appears in the Epstein flight logs ${passenger.flightCount} times ${range(
    passenger.histogram,
  )}`;
};

export default ({ match }) => {
  const { slug } = match.params;
  const { loading, error, data, fetchMore } = useQuery(PASSENGER, {
    variables: { slug, limit: FLIGHT_LIMIT },
  });

  if (loading) return <Loading text />;
  if (error) return <p>{error.message}</p>;

  return (
    <>
      <MetaTags
        title={data.passenger.name}
        uri={`passenger/${data.passenger.slug}`}
        description={`${data.passenger.name} ${historyText(data).replace(
          /(^[A-Z])/,
          cap => cap.toLowerCase(),
        )}`}
      />
      <Details>
        <span>name</span>
        <span>{data.passenger.name}</span>
        <span>bio</span>
        <div>
          <span>{data.passenger.biography}</span>
          <br />
          <a className={link} href={data.passenger.wikipedia_link}>
            <span>See more on Wikipedia â†’</span>
          </a>
        </div>
        <span>history</span>
        <div>{historyText(data)}</div>
        {data.passenger.image ? (
          <>
            <span>image</span>
            <img src={data.passenger.image} alt="" />
          </>
        ) : null}
        <span>flights</span>
        <PaginatedBrowser
          browserComponent={MiniFlight}
          ids={data.passenger.flights.edges.map(x => x.id)}
          totalAvailable={data.passenger.flightCount}
          pageSize={FLIGHT_LIMIT}
          fetchMore={() =>
            fetchMore({
              variables: { offset: data.passenger.flights.edges.length + 1 },
              updateQuery: (prev, { fetchMoreResult }) => {
                if (!fetchMoreResult) return prev;
                return {
                  passenger: {
                    ...prev.passenger,
                    flights: {
                      ...fetchMoreResult.passenger.flights,
                      pageInfo: {
                        ...fetchMoreResult.passenger.flights.pageInfo,
                        count:
                          prev.passenger.flights.pageInfo.count +
                          fetchMoreResult.passenger.flights.pageInfo.count,
                      },
                      edges: [
                        ...prev.passenger.flights.edges,
                        ...fetchMoreResult.passenger.flights.edges,
                      ],
                    },
                  },
                };
              },
            })
          } // eslint-disable-line
        />
      </Details>
    </>
  );
};
