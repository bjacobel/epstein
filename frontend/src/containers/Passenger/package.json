{
  "main": "./Passenger.js",
  "private": true
}
