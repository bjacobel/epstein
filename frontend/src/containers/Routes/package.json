{
  "main": "./Routes.js",
  "private": true
}
