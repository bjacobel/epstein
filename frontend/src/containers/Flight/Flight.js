import React, { Suspense } from 'react';
import { useQuery, gql } from '@apollo/client';
import { format, parseISO } from 'date-fns';
import { Link } from 'react-router-dom';

import Details from '../../components/Details';
import Loading from '../../components/Loading';
import MetaTags from '../../components/MetaTags';
import { passengers, noslug, explainLiteral } from './style.css';
import { link } from '../../stylesheets/shared.css';

export const FLIGHT = gql`
  fragment airfield on Airfield {
    name
    iata_code
    latitude_deg
    longitude_deg
    municipality
    iso_country
  }
  query Flight($id: Int!) {
    flight(id: $id) {
      id
      distance
      date
      page
      passengers {
        edges {
          ... on VerifiedPassenger {
            name
            slug
          }
          ... on LiteralPassenger {
            literal
          }
        }
      }
      aircraft {
        id
        model
        tailsign
      }
      source {
        ...airfield
      }
      destination {
        ...airfield
      }
    }
  }
`;

const formatKm = m =>
  `${new Intl.NumberFormat().format(Number.parseFloat(m / 1000).toPrecision(4))} km`;

/* eslint-disable camelcase */
const Airfield = ({ name, municipality, iso_country }) => (
  <span>{`${name} (${municipality}, ${iso_country})`}</span>
);
/* eslint-enable camelcase */

export const describePassengers = plist => {
  const notablePassengers = plist.filter(x => x.name).map(x => x.name);
  const listablePassengers = notablePassengers.slice(0, notablePassengers.length - 1);
  const listedPassengers = listablePassengers.join(', ');
  const lastPassenger = notablePassengers[notablePassengers.length - 1];

  return `${plist.length} passengers were listed on the manifest${
    notablePassengers.length
      ? `, including ${listedPassengers}${
          listedPassengers.length ? ' and ' : ''
        }${lastPassenger}`
      : ''
  }`;
};

export default ({ match }) => {
  const { id } = match.params;
  const { loading, error, data } = useQuery(FLIGHT, { variables: { id } });

  if (loading) return <Loading text />;
  if (error) throw error;

  const isoDate = parseISO(data.flight.date);

  const SuspenseMap = React.lazy(() =>
    import(/* webpackChunkName: "map" */ '../../components/Map'),
  );

  return (
    <>
      <MetaTags
        title={`${data.flight.source.iata_code} to ${
          data.flight.destination.iata_code
        } on ${format(isoDate, 'M/d/yy')}`}
        description={describePassengers(data.flight.passengers.edges)}
        uri={`flight/${data.flight.id}`}
      />
      <Details>
        <span>date</span>
        <span>{format(isoDate, 'MMM d, y')}</span>
        <span>source</span>
        <Airfield {...data.flight.source} />
        <span>destination</span>
        <Airfield {...data.flight.destination} />
        <span>distance</span>
        <span>{formatKm(data.flight.distance)}</span>
        <span>map</span>
        <div>
          <Suspense fallback={<Loading />}>
            <SuspenseMap
              source={[data.flight.source.latitude_deg, data.flight.source.longitude_deg]}
              dest={[
                data.flight.destination.latitude_deg,
                data.flight.destination.longitude_deg,
              ]}
            />
          </Suspense>
        </div>
        <span>source</span>
        <span>
          <a
            className={link}
            href={data.flight.page}
            target="_blank"
            rel="noopener noreferrer"
          >
            original PDF document
          </a>
        </span>
        <span>passengers</span>
        <div className={passengers}>
          <ul>
            {data.flight.passengers.edges
              .filter(x => x.slug)
              .map(({ slug, name }) => (
                <li key={slug}>
                  <Link className={link} to={`/passenger/${slug}`}>
                    {name}
                  </Link>
                </li>
              ))}
            <hr />
            <span className={explainLiteral}>
              <span>As well as the following, who </span>
              <a
                className={link}
                href="/about#why-are-some-entries-not-linked-to-an-identified-passenger"
              >
                have not been linked to an identified passenger
              </a>
              :
            </span>
            {data.flight.passengers.edges
              .filter(x => !x.slug)
              .map(({ literal }) => (
                <li key={literal} className={noslug}>
                  <span>{`“${literal}”`}</span>
                </li>
              ))}
          </ul>
        </div>
        <span>aircraft</span>
        <div>
          <span>{data.flight.aircraft.model}</span>
          <br />
          <span>
            <span>tailsign </span>
            <Link className={link} to={`/aircraft/${data.flight.aircraft.id}`}>
              {data.flight.aircraft.tailsign}
            </Link>
          </span>
        </div>
      </Details>
    </>
  );
};
