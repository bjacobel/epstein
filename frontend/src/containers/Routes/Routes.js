import React, { Component, Suspense } from 'react';
import { Route, Switch, Redirect } from 'react-router-dom';

import NotFound from '../NotFound';
import Home from '../Home';
import Flight from '../Flight';
import Passenger from '../Passenger';
import Login from '../Login';
import Search from '../Search';
import Analytics from '../../utils/Analytics';
import Loading from '../../components/Loading';
import { hasValidJwtToken } from '../../utils/auth';

export class GARoute extends Route {
  render() {
    this.props.ga.pageview();
    // eslint-disable-next-line react/jsx-props-no-spreading
    return <Route {...this.props} />;
  }
}

const SuspenseAdmin = React.lazy(() =>
  import(/* webpackChunkName: "admin" */ '../Admin'),
);

const SuspenseAbout = React.lazy(() =>
  import(/* webpackChunkName: "about" */ '../About'),
);

export const PrivateRoute = ({ children, ...rest }) => (
  <Route
    {...rest}
    render={({ location }) =>
      hasValidJwtToken() ? (
        children
      ) : (
        <Redirect
          to={{
            pathname: '/login',
            state: { from: location },
          }}
        />
      )
    }
  />
);

export default class Routes extends Component {
  constructor() {
    super();
    this.ga = new Analytics();
  }

  render() {
    return (
      <div>
        <Switch>
          <GARoute ga={this.ga} path="/" exact component={Home} />
          <GARoute ga={this.ga} path="/flight/:id" exact component={Flight} />
          <GARoute ga={this.ga} path="/passenger/:slug" exact component={Passenger} />
          <GARoute
            ga={this.ga}
            path="/about"
            exact
            render={() => (
              <Suspense fallback={<Loading />}>
                <SuspenseAbout />
              </Suspense>
            )}
          />
          <Route path="/login" exact component={Login} />
          <PrivateRoute path="/admin" exact>
            <Suspense fallback={<Loading />}>
              <SuspenseAdmin />
            </Suspense>
          </PrivateRoute>
          <GARoute ga={this.ga} path="/search/:query?" exact component={Search} />
          <GARoute ga={this.ga} component={NotFound} />
        </Switch>
      </div>
    );
  }
}
