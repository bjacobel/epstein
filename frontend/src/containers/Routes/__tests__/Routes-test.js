import React from 'react';
import { shallow, mount } from 'enzyme';
import { BrowserRouter, Route, Switch } from 'react-router-dom';

import Routes, { PrivateRoute } from '../Routes';
import Flight from '../../Flight';
import Passenger from '../../Passenger';
import Analytics from '../../../utils/Analytics';
import { hasValidJwtToken } from '../../../utils/auth';

jest.mock('containers/Home', () => () => null);
jest.mock('containers/Flight', () => () => null);
jest.mock('containers/Passenger', () => () => null);
jest.mock('containers/NotFound');
jest.mock('utils/Analytics');
jest.mock('utils/auth');

const setPath = value => {
  global.history.pushState({}, value, value);
};

describe('Router', () => {
  let RoutedRoutes;
  beforeEach(() => {
    RoutedRoutes = () => (
      <BrowserRouter>
        <Routes />
      </BrowserRouter>
    );
  });

  it('matches snapshot', () => {
    expect(shallow(<Routes />)).toMatchSnapshot();
  });

  describe('Routing config', () => {
    it('routes flights/:id to the Flight component', () => {
      setPath('/flight/125');
      const routes = mount(<RoutedRoutes />);
      expect(routes.find(Flight).length).toBe(1);
      expect(routes.find(Flight).props()).toEqual(
        expect.objectContaining({
          match: expect.objectContaining({
            params: {
              id: '125',
            },
          }),
        }),
      );
    });

    it('routes passenger/:slug to the Passenger component', () => {
      setPath('/passenger/bill-clinton');
      const routes = mount(<RoutedRoutes />);
      expect(routes.find(Passenger).length).toBe(1);
      expect(routes.find(Passenger).props()).toEqual(
        expect.objectContaining({
          match: expect.objectContaining({
            params: {
              slug: 'bill-clinton',
            },
          }),
        }),
      );
    });

    it('has a fallthrough 404', () => {
      setPath('/asdfasaddf');
      const routes = mount(<RoutedRoutes />);
      expect(routes.find('NotFound').length).toBe(1);
    });
  });

  describe('Analytics functionality', () => {
    beforeEach(() => {
      // reset call count before each test
      Analytics.prototype.pageview.mockClear();
    });

    it('inits the analytics class', () => {
      mount(<RoutedRoutes />);
      expect(Analytics).toHaveBeenCalled();
    });

    it('calls the analytics pageview fn when you navigate to a new route', () => {
      mount(<RoutedRoutes />);
      setPath('/child/1');
      expect(Analytics.prototype.pageview).toHaveBeenCalledTimes(1);
    });

    it('supplies GA to routed components via props', () => {
      const routes = mount(<RoutedRoutes />);
      expect(routes.find('Route').props()).toEqual(
        expect.objectContaining({
          ga: {},
        }),
      );
    });
  });

  describe('private routes', () => {
    let PrivateRoutes;
    beforeEach(() => {
      PrivateRoutes = () => (
        <BrowserRouter>
          <Switch>
            <Route exact path="/login" render={() => 'login page'} />
            <PrivateRoute>
              <Route exact path="/admin" render={() => 'protected resource'} />
            </PrivateRoute>
          </Switch>
        </BrowserRouter>
      );
    });

    it('redirects when auth is not present', () => {
      hasValidJwtToken.mockReturnValueOnce(false);
      setPath('/admin');
      const wrapper = mount(<PrivateRoutes />);
      expect(wrapper.text()).toEqual('login page');
    });

    it('renders the specified child when auth is present', () => {
      hasValidJwtToken.mockReturnValueOnce(true);
      setPath('/admin');
      const wrapper = mount(<PrivateRoutes />);
      expect(wrapper.text()).toEqual('protected resource');
    });
  });
});
