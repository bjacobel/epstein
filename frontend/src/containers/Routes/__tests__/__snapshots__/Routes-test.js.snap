// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Router matches snapshot 1`] = `
<div>
  <Switch>
    <GARoute
      component={[MockFunction]}
      exact={true}
      ga={Object {}}
      path="/"
    />
    <GARoute
      component={[Function]}
      exact={true}
      ga={Object {}}
      path="/flight/:id"
    />
    <GARoute
      component={[Function]}
      exact={true}
      ga={Object {}}
      path="/passenger/:slug"
    />
    <Route
      component={[Function]}
      exact={true}
      path="/login"
    />
    <PrivateRoute
      exact={true}
      path="/admin"
    >
      <Suspense
        fallback={<_default />}
      >
        <lazy />
      </Suspense>
    </PrivateRoute>
    <GARoute
      component={[MockFunction]}
      ga={Object {}}
    />
  </Switch>
</div>
`;
